pkg/controllers/deprovisioning/suite_test.go

```diff
@@ -130,6 +131,259 @@ var _ = AfterEach(func() {
+			ExpectTriggerVerifyAction(&wg)
+			ExpectReconcileFailed(ctx, deprovisioningController, client.ObjectKey{})
+		}()
+
+		// Iterate in a loop until we get to the validation action
+		// Then, apply the pods to the cluster and bind them to the nodes
+		for {
+			time.Sleep(100 * time.Millisecond)
+			if len(ExpectNodeClaims(ctx, env.Client)) == 2 {
+				break
+			}
+		}
+		nodeClaimNode = ExpectNodeExists(ctx, env.Client, nodeClaimNode.Name)
+		Expect(nodeClaimNode.Spec.Taints).To(ContainElement(v1beta1.DisruptionNoScheduleTaint))
+
+		createdNodeClaim := lo.Reject(ExpectNodeClaims(ctx, env.Client), func(nc *v1beta1.NodeClaim, _ int) bool {
+			return nc.Name == nodeClaim.Name
+		})
+		ExpectDeleted(ctx, env.Client, createdNodeClaim[0])
+		ExpectNodeClaimsCascadeDeletion(ctx, env.Client, createdNodeClaim[0])
+		ExpectNotFound(ctx, env.Client, createdNodeClaim[0])
+
+		wg.Wait()
+		nodeClaimNode = ExpectNodeExists(ctx, env.Client, nodeClaimNode.Name)
+		Expect(nodeClaimNode.Spec.Taints).ToNot(ContainElement(v1beta1.DisruptionNoScheduleTaint))
+	})
+	It("should remove taints from Machines that fail to deprovision", func() {
+		provisioner.Spec.Consolidation = &v1alpha5.Consolidation{Enabled: lo.ToPtr(true)}
+		pod := test.Pod(test.PodOptions{
+			ResourceRequirements: v1.ResourceRequirements{
+				Requests: v1.ResourceList{
+					v1.ResourceCPU:    resource.MustParse("100m"),
+					v1.ResourceMemory: resource.MustParse("100Mi"),
+				},
+			},
+		})
+		ExpectApplied(ctx, env.Client, provisioner, machine, machineNode, pod)
+		ExpectManualBinding(ctx, env.Client, pod, machineNode)
+
+		// inform cluster state about nodes and machines
+		ExpectMakeNodesAndMachinesInitializedAndStateUpdated(ctx, env.Client, nodeStateController, machineStateController, []*v1.Node{machineNode}, []*v1alpha5.Machine{machine})
+
+		fakeClock.Step(10 * time.Minute)
+
+		// Trigger the reconcile loop to start but don't trigger the verify action
+		wg := sync.WaitGroup{}
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			ExpectTriggerVerifyAction(&wg)
+			ExpectReconcileFailed(ctx, deprovisioningController, client.ObjectKey{})
+		}()
+
+		// Iterate in a loop until we get to the validation action
+		// Then, apply the pods to the cluster and bind them to the nodes
+		for {
+			time.Sleep(100 * time.Millisecond)
+			if len(ExpectMachines(ctx, env.Client)) == 2 {
+				break
+			}
+		}
+		machineNode = ExpectNodeExists(ctx, env.Client, machineNode.Name)
+		Expect(machineNode.Spec.Unschedulable).To(BeTrue())
+
+		createdMachine := lo.Reject(ExpectMachines(ctx, env.Client), func(m *v1alpha5.Machine, _ int) bool {
+			return m.Name == machine.Name
+		})
+		ExpectDeleted(ctx, env.Client, createdMachine[0])
+		ExpectMachinesCascadeDeletion(ctx, env.Client, createdMachine[0])
+		ExpectNotFound(ctx, env.Client, createdMachine[0])
+		wg.Wait()
+		machineNode = ExpectNodeExists(ctx, env.Client, machineNode.Name)
+		Expect(machineNode.Spec.Unschedulable).To(BeFalse())
+	})
+})
+
```

Comment:
Do we have a test case to validate that we remove the taints on re-reconcile if they are left behind after the original reconcile loop? You can probably just mock this behavior by adding the taint to a NodeClaim node and then just executing a reconcile to ensure that it gets removed